consumer.go

```diff
@@ -12,58 +12,97 @@ import (
 	"github.com/Shopify/sarama"
 )
 
-const errConsumerNotConnected = "consumer is not connected to Kafka"
+var (
+	errConsumerNotConnected = fmt.Errorf("consumer is not connected to Kafka")
+	errMonitorNotConnected  = fmt.Errorf("consumer monitor is not connected to Kafka")
+)
 
 // Consumer which will keep trying to reconnect to Kafka on a specified interval.
 // The underlying consumer group is created lazily when message listening is started.
 type Consumer struct {
-	config            ConsumerConfig
-	consumerGroupLock *sync.RWMutex
-	consumerGroup     sarama.ConsumerGroup
-	retryInterval     time.Duration
-	logger            *logger.UPPLogger
-	closed            chan struct{}
+	config                  ConsumerConfig
+	topics                  []*Topic
+	consumerGroupLock       *sync.RWMutex
+	consumerGroup           sarama.ConsumerGroup
+	monitorLock             *sync.RWMutex
+	monitor                 *consumerMonitor
+	connectionRetryInterval time.Duration
+	logger                  *logger.UPPLogger
+	closed                  chan struct{}
 }
 
 type ConsumerConfig struct {
 	BrokersConnectionString string
 	ConsumerGroup           string
-	Topics                  []string
-	Options                 *sarama.Config
+	// Time interval between each offset fetching request. Default is 3 minutes.
+	OffsetFetchInterval time.Duration
+	// Total count of offset fetching request failures until consumer status is marked as unknown.
+	// Default is 5. Note: A single failure will result in follow-up requests to be sent on
+	// shorter interval than the value of OffsetFetchInterval until successful.
+	OffsetFetchMaxFailureCount int
+	Options                    *sarama.Config
 }
 
-func NewConsumer(config ConsumerConfig, log *logger.UPPLogger, retryInterval time.Duration) *Consumer {
+func NewConsumer(config ConsumerConfig, topics []*Topic, log *logger.UPPLogger, connectionRetryInterval time.Duration) *Consumer {
 	consumer := &Consumer{
-		config:            config,
-		consumerGroupLock: &sync.RWMutex{},
-		retryInterval:     retryInterval,
-		logger:            log,
-		closed:            make(chan struct{}),
+		config:                  config,
+		topics:                  topics,
+		consumerGroupLock:       &sync.RWMutex{},
+		monitorLock:             &sync.RWMutex{},
+		connectionRetryInterval: connectionRetryInterval,
+		logger:                  log,
+		closed:                  make(chan struct{}),
 	}
+
 	return consumer
 }
 
 // connect will attempt to create a new consumer continuously until successful.
 func (c *Consumer) connect() {
```

Comment:
That's actually a great point and can be investigated for a future release. Right now the producer creation is also done like that and I could argue it's actually worse there because the `connect()` function is ran inside a go routine and its lifecycle is not monitored.

We can come up with an API design where the connections **must** be established when a service is being initialized and if they fail, the service would exit with an error and be restarted by k8s until connection is successful.

I believe in such scenario we'd be able to omit the one-time synchronization usages we're currently having. It'd be great to do this with the upcoming message format changes in a new major version.